src/radical/pilot/utils/serializer.py

```diff
@@ -0,0 +1,154 @@
+
+"""
+The serializer should be able to (de)serialize information that we want
+to send over the wire from the client side to the agent side via
+1- ZMQ
+2- MongoDB
+
+we except:
+    1- Callables with and without dependecies.
+    2- Non-callables like classes and other python objects
+"""
+__copyright__ = "Copyright 2013-2016, http://radical.rutgers.edu"
+__license__   = "MIT"
+
+import os
+import dill
+import pickle
+import codecs
+import tempfile
+
+
+_obj_dir       = tempfile.gettempdir()
+_obj_file_name = 'rp_obj.pkl'
+_obj_file_path = os.path.join(_obj_dir, _obj_file_name)
+
+
+def serialize_obj(obj):
+    """
+    serialize object
+    """
+    result    = None
+    exception = None
+
+    if callable(obj):
+        try:
+            result = dill.dumps(obj)
+        except Exception as e:
+            exception = e
+            pass
+
+        # if we fail, then pikle it by reference
+        if result is None:
```

Comment:
> we do not want the serializer to stop on one way of serialization (in case it fails). We continue to try to serialize with different approaches until all of them fail

@AymenFJA in case your comment was related to this part, here is what I meant:
```python
    if callable(obj):
        try:
            result = dill.dumps(obj)
        except Exception:
            try:
                result = dill.dumps(obj, byref = True)
            except Exception as e:
                pass
```